webR 0.2.0 release post (#652)
tidyverse/tidyverse.org: content/blog/webr-0-2-0/index.Rmd

categories: [deep-dive]
tags: [webr, webassembly, wasm]
---

```{=html}
<!--
TODO:
* [x] Look over / edit the post's title in the yaml
* [x] Edit (or delete) the description; note this appears in the Twitter card
* [x] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
* [x] Find photo & update yaml metadata
* [x] Create `thumbnail-sq.jpg`; height and width should be equal
* [x] Create `thumbnail-wd.jpg`; width should be >5x height
* [x] `hugodown::use_tidy_thumbnails()`
* [x] Add intro sentence, e.g. the standard tagline for the package
* [x] `usethis::use_tidy_thanks()`
* [] Release webR 0.2.0
* [] Update all links from /0.2.0-rc.1 to /0.2.0
* [] Update webr-repo packages
* [] Update webr-repo dashboard
-->
```
<!-- Initialise webR in the page -->

`r hugodown::shortcode("webr-init")`

<!-- Add webr engine for knit -->

```{r, echo=FALSE}
knitr::knit_engines$set(webr = function(options) {
  code <- paste(options$code, collapse = "\n")
  w <- knitr::opts_current$get('fig.width') * 72
  h <- knitr::opts_current$get('fig.height') * 72
  hugodown::shortcode("webr-editor", code = I(encodeString(code, quote = '`')), width = w, height = h)
})
```

> We're absolutely thrilled to announce the release of webR 0.2.0! This release gathers together many updates and improvements to webR over the last few months, including improvements to the HTML canvas graphics device, support for Cario-based bitmap graphics, accessibility and internationalisation improvements, additional Wasm R package support (including Shiny), a new webR REPL app, and various updates to the webR developer API.

Reviewer: I would link to webR docs page from the first mention.

[^2]: Also other JavaScript/Wasm environments, such as Node.js. For example, [ROpenSci's r-universe](https://ropensci.org/r-universe/) package platform provides download links for datasets contained in R packages, in a variety of formats, [powered by running webR server-side in Node.js](https://fosstodon.org/@jeroenooms/110299179903212170).

Running R code directly in a browser opens the door for many new and exciting uses for R on the web. Applications that I'm personally excited in seeing developed are,

-   Live and interactive R code and graphics in documents & presentations,
-   Tactile educational content for R, with examples that can be remixed on-the-fly by learners,
-   Reproducible statistics through containerisation and notebook-style literate programming.

Even in these early days, some of this is already being provided by development of downstream projects such as James Balamuta's [quarto-webr](https://github.com/coatless/quarto-webr) extension, allowing Quarto users to easily embed interactive R code blocks in their documents.

### Interactive code blocks

> One of my favourite demonstrations of what webR can do is interactive code blocks for R code. After a short loading period while the webR binary is downloaded, a **Run code** button will be enabled below. Using examples like this, R code can be immediately edited and executed -- feel free to experiment!

Reviewer: You might add this to the end to encourage experimentation.

> Click the "Run code" button to see the resulting box plot, change the color from `mediumseagreen` to `red` and run the code again.

Author: Good idea!

```{webr}
colnames(mtcars)

boxplot(mpg ~ cyl, data = mtcars, col = "mediumseagreen",
        xlab = "Number of Cylinders", ylab = "Miles/(US) gallon", 
        main = "Motor Trend Car Road Tests",
        sub = "Source: 1974 Motor Trend US magazine")
```

It's easy to see the potential teaching benefit examples like this could bring to educational content or R package documentation.

## The webR REPL app

> WebR can be loaded into a web page to be used as a part of a wider web application, and ships with a demo application that does just that. The webR REPL app[^3] provides a simple R environment directly in your web browser. The app can be accessed at the URL https://webr.r-wasm.org/v0.2.0-rc.1/ and included are sections for R console input/output, code editing, file management, and graphics device output. 

Reviewer: ```suggestion
WebR can be loaded into a web page to be used as a part of a wider web application, and ships with a demo application that does just that. The webR REPL app[^3] provides a simple R environment directly in your web browser. The app can be accessed at the URL https://webr.r-wasm.org/v0.2.0-rc.1/ and includes sections for R console input/output, code editing, file management, and graphics device output. 
```

[^3]: REPL stands for "Read, Eval, Print, Loop", and is another name for the R console that you're probably familiar with. The application is named the "webR REPL app" because the original version simply provided the user with a fullscreen R console in their web browser.

> With the webR REPL app a casual user could get up and running with R in seconds, without having to install any software to their machine. It is entirely feasible that they could perform the basics of data science entirely within their web browser!

Reviewer: ```suggestion
With the webR REPL app, a casual user could get up and running with R in seconds, without having to install any software on their machine. It is entirely feasible that they could perform the basics of data science entirely within their web browser!
```

Other than interactive code blocks, like in the example earlier, the webR REPL app is perhaps the first thing that users new to webR will interact with. For this reason, we have spent some time working to improve the technical implementation and user experience of using the app. The app has been completely rewritten in the React web framework, replacing the older jQuery library. This allows for better component code organisation and more rapid development of features and updates.

<a href="repl.png"><img alt="A screenshot the webR REPL app. The code to generate a ggplot, along with its output, is shown in the app." width="95%" src="repl.png"></a>

### Code editor

> The app now comes with a tabbed code editor, allowing for easier editing and execution of R code. The editor integrates with the webR virtual filesystem (VFS), meaning that multiple R scripts can be opened, edited and saved and they will be available to the running Wasm R process.

Reviewer: ```suggestion
The app now comes with a tabbed code editor, allowing for easier editing and execution of R code. The editor integrates with the webR virtual filesystem (VFS), meaning that multiple R scripts can be opened, edited, and saved and they will be available to the running Wasm R process.
```

The editor is integrated with the currently running R process and automatic code suggestions are shown as you type, provided by R's [built in completion generator](https://stat.ethz.ch/R-manual/R-devel/library/utils/html/rcompgen.html). The suggestions are context sensitive and are aware of package and function names, valid arguments, and even objects that exist in the global environment.

<a href="completion.png"><img alt="A screenshot of the editor component showing code completion results. One of the suggestions is a data set available in the global environment." width="70%" src="completion.png"></a>

The running Wasm R process is also configured at initialisation to use the editor component as its display [pager mechanism](https://stat.ethz.ch/R-manual/R-devel/library/base/html/file.show.html). With this configuration in place running commands such as `?rnorm` in the app automatically opens a new read-only tab in the editor displaying R's built-in documentation.

<a href="documentation.png"><img alt="A screenshot of the editor component showing built-in R documentation" width="80%" src="documentation.png"></a>

### Plotting pane

> The plotting pane has been updated to take advantage of improvements in webR's HTML canvas graphic device, set as the default device as part of initialisation. In particular, multiple plots are now supported and older plots can be directly accessed using the previous and next buttons in the plotting toolbar. You can try this out with R's built in graphics demo, by running `demo(graphics)` and/or `demo(persp)`.

Reviewer: Is it "graphic device" or "graphics device"? I'm not sure, but I feel like I see the latter more often, at least in other contexts.

Author: I was aiming for "graphics device", this was a typo. Thanks!

<a href="plotting.png"><img alt="A screenshot of the plot pane showing a built-in R graphics demo" width="75%" src="plotting.png"></a>

### Files pane

> The files pane has been completely redesigned, removing is dependency on jQuery and instead making use of the [react-accessible-treeview](https://www.npmjs.com/package/react-accessible-treeview) package. As well as a technical improvement, this change means that interacting with the webR filesystem should be more usable to those with web accessibility requirements. We feel it's important that, where possible, everybody is able to use our software.

Reviewer: ```suggestion
The files pane has been completely redesigned, removing its dependency on jQuery and instead making use of the [react-accessible-treeview](https://www.npmjs.com/package/react-accessible-treeview) package. As well as a technical improvement, this change means that interacting with the webR filesystem should be more usable to those with web accessibility requirements. We feel it's important that, where possible, everybody is able to use our software.
```

In comparison, web browser support for text rendering is excellent. Now that we use the `OffscreenCanvas` interface, we too can take advantage of the years of work behind browser's support for text on the web. The example below demonstrates several of the modern text rendering features now supported by [`webr::canvas()`](https://docs.r-wasm.org/webr/latest/api/r.html#graphics-device-for-drawing-to-a-html-canvas-element).

```{webr}
plot(rnorm(1000), rnorm(1000), col=rgb(0,0,0,0.5),
     xlim=c(-5, 5), ylim=c(-5, 5), main="This is the title 🚀",
     xlab="This is the x label", ylab="This is the y label",
     family="Futura")
text(-3.5, 4, "This is English", family="monospace")
text(-3.5, -4, "هذا مكتوب باللغة العربية")
text(3.5, 4, "これは日本語です")
text(3.5, -4, "זה כתוב בעברית")

> ```

Reviewer: In code blocks, I recommend spaces around `=` to conform to the Tidyverse style guide. I would also recommend a line break after the first `(`, something like

```r
plot(
  rnorm(1000), rnorm(1000), 
  col = rgb(0,0,0,0.5),
  xlim = c(-5, 5), ylim = c(-5, 5), 
  main = "This is the title 🚀",
  xlab = "This is the x label", 
  ylab = "This is the y label",
  family = "Futura"
)
text(-3.5, 4, "This is English", family="monospace")
text(-3.5, -4, "هذا مكتوب باللغة العربية")
text(3.5, 4, "これは日本語です")
text(3.5, -4, "זה כתוב בעברית")
```

[^6]: [Dyslexie](https://www.dyslexiefont.com), [Open Dyslexic](https://opendyslexic.org). Results of research in this area is mixed, but even if these fonts don't improve the speed of text comprehension, some users may simply prefer or feel more comfortable with them.

Font metrics are now exact, using [`measureText()`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/measureText), rather than estimating the width and height of Latin glyphs using heuristics. This gives more accurate positioning of rendered text and improves the general quality of resulting plots.

Support for Unicode, font glyph fallback, complex ligatures, and right-to-left (RTL) text have all been improved. This vastly improves results when rendering text for international users, particularly for non-Latin RTL scripts such as the Arabic and Hebrew text in the example above.

> Also, colour emoji can now be added to plots 😃

Reviewer: ```suggestion
Also, colour emoji can now be added to plots. 😃
```

Consider the example below. Data is transferred from JavaScript into the running R process by binding `jsData` to an R variable in the global environment using [`webR.objs.globalEnv.bind()`](https://docs.r-wasm.org/webr/latest/convert-js-to-r.html#binding-objects-to-an-r-environment). Next, some computation on the data is done, represented as evaluating the `do_analysis()` R function. Finally the result is returned back to JavaScript, first as a reference to an R object and then transferring the result data back to the JavaScript environment using [`toJs()`](https://docs.r-wasm.org/webr/latest/convert-r-to-js.html#serialising-r-objects).

``` javascript
const jsData = [... some large JavaScript dataset ...];
await webR.objs.globalEnv.bind('data', jsData);

const ret = await webR.evalR("do_analysis(data)");
const result = await ret.toJs();
```

It's easy to see how this workflow could be useful as part of a wider application, enabling a complex data manipulation or a statistical modelling in R that would otherwise be awkward to perform directly in JavaScript.

> Behind the scenes we've done work to ensure that data is transferred efficiently to and from the R environment, and in webR 0.2.0 the [MessagePack](https://msgpack.org/index.html) protocol is now used as the main way that data is serialised and transferred, replacing JSON encoding.

Reviewer: ```suggestion
Behind the scenes, we've done work to ensure that data is transferred efficiently to and from the R environment, and in webR 0.2.0 the [MessagePack](https://msgpack.org/index.html) protocol is now used as the main way that data is serialised and transferred, replacing JSON encoding.
```

A coming-soon version of Shinylive for R will provide a much better user experience for getting fully client-side R Shiny apps up and running, without requiring advanced knowledge of JavaScript's Worker API. I believe Shinylive with webR integration will pave the way for providing a user-friendly method to build and deploy containerised R Shiny apps, running on WebAssembly.

## Installation and next steps

Developers can integrate webR in their own JavaScript or TypeScript projects by installing the [webR npm package](https://www.npmjs.com/package/webr), or by directly importing webR from CDN. Issues and PRs are accepted and welcome on the main [r-wasm/webr](https://github.com/r-wasm/webr) GitHub repository.

### npm

With this release, the webR npm package name has been updated, simplified from the original `@r-wasm/webr` package name to simply `webr`.

``` bash
npm i webr
```

The original namespaced package `@r-wasm/webr` will be deprecated, and from v0.2.0 onwards npm will display a message pointing to the new package name.

### CDN URL

> Alternatively, webR can be imported directly as a module from CDN,

Reviewer: ```suggestion
Alternatively, webR can be imported directly as a module from CDN.
```

* [x] Add intro sentence, e.g. the standard tagline for the package
* [x] `usethis::use_tidy_thanks()`
* [x] Release webR 0.2.0
* [] Update all links from /0.2.0-rc.1 to /0.2.0
* [x] Update webr-repo packages
* [x] Update webr-repo dashboard
-->
```
<!-- Initialise webR in the page -->

`r hugodown::shortcode("webr-init")`

<!-- Add webr engine for knit -->

```{r, echo=FALSE}
knitr::knit_engines$set(webr = function(options) {
  code <- paste(options$code, collapse = "\n")
  w <- knitr::opts_current$get('fig.width') * 72
  h <- knitr::opts_current$get('fig.height') * 72
  hugodown::shortcode("webr-editor", code = I(encodeString(code, quote = '`')), width = w, height = h)
})
```

We're absolutely thrilled to announce the release of [webR](https://docs.r-wasm.org/webr/v0.2.0/) 0.2.0! This release gathers together many updates and improvements to webR over the last few months, including improvements to the HTML canvas graphics device, support for Cario-based bitmap graphics, accessibility and internationalisation improvements, additional Wasm R package support (including Shiny), a new webR REPL app, and various updates to the webR developer API.

> This will be a fairly technical blog post, taking a deep dive through the major breaking changes and new features available in webR 0.2.0. I also plan to record and release a series of companion videos discussing the new release, so keep an eye out if you're someone who prefers watching and listening over reading long-form articles. I'll update this post with all the links once they're available.

Author: Fair enough, I'll rephrase without it :)
